rustls/examples/internal/bogo_shim.rs

```diff
@@ -226,6 +251,28 @@ impl client::ServerCertVerifier for DummyServerAuth {
     ) -> Result<client::ServerCertVerified, Error> {
         Ok(client::ServerCertVerified::assertion())
     }
+
+    fn verify_tls12_signature(
+        &self,
+        message: &[u8],
+        cert: &rustls::Certificate,
+        dss: &DigitallySignedStruct,
+    ) -> Result<HandshakeSignatureValid, Error> {
+        WebPkiServerVerifier::default_verify_tls12_signature(message, cert, dss)
```

Comment:
I wonder if there's a way we can factor this further to avoid the duplication here? We have identical `verify_tls12_signature()` and `verify_tls13_signature()` methods for both `ClientCertVerifier` and `ServerCertVerifier` and we end up with the `ClientCertVerifier` methods calling into the `ServerCertVerifier` methods to work around that.

The signature verification is different from other parts of the certificate verification because (a) no verifier state is involved and (b) it's the same across client and server. Feels like it might make sense to extract a separate trait for this?

Reply:
Yeah this is a bit of a separate concern from certificate chain building, and it's mainly like this because it draws a line between things that require x509 certificate parsing and everything else, and I'm not sure it makes sense to have two separate configuration surfaces for "stuff that needs to deal with certificates"?

In future commits I'm planning to introduce some shared configuration between these which is the [signature verification schemes to be used by webpki](https://github.com/rustls/rustls/blob/4a4d61b5534328254f02110be6ef582a13551bbc/rustls/src/webpki/verify.rs#L480-L504) -- which would be something shared by a certificate verifier for chain building and signature verification (though, still, the configuration is separate for both, but I think it makes sense for that to be declared together).